fiftyone/core/odm/utils.py

```diff
@@ -0,0 +1,159 @@
+        value: a value
+
+    Returns:
+        a field specification dict
+    """
+    if isinstance(value, foed.BaseEmbeddedDocument):
+        return {
+            "ftype": fof.EmbeddedDocumentField,
+            "embedded_doc_type": type(value),
+            "fields": get_embedded_document_fields(value),
+        }
+
+    if isinstance(value, bool):
+        return {"ftype": fof.BooleanField}
+
+    if isinstance(value, numbers.Integral):
+        return {"ftype": fof.IntField}
+
+    if isinstance(value, numbers.Number):
+        return {"ftype": fof.FloatField}
+
+    if isinstance(value, six.string_types):
+        return {"ftype": fof.StringField}
+
+    if isinstance(value, datetime):
+        return {"ftype": fof.DateTimeField}
+
+    if isinstance(value, date):
+        return {"ftype": fof.DateField}
+
+    if isinstance(value, (list, tuple)):
+        kwargs = {"ftype": fof.ListField}
+
+        value_types = set(_get_list_value_type(v) for v in value)
+
+        if value_types == {fof.IntField, fof.FloatField}:
+            kwargs["subfield"] = fof.FloatField
+        elif len(value_types) == 1:
+            value_type = next(iter(value_types))
+            if value_type is not None:
+                kwargs["subfield"] = value_type
+
+            if value_type == fof.EmbeddedDocumentField:
+                document_types = {v.__class__ for v in value}
+                if len(document_types) > 1:
+                    raise ValueError("Cannot merge types")
+
+                kwargs["embedded_doc_type"] = document_types.pop()
+
+                kwargs["fields"] = _merge_implied_fields(
+                    [get_implied_field_kwargs(v) for v in value]
+                )
+
+        return kwargs
+
+    if isinstance(value, np.ndarray):
+        if value.ndim == 1:
+            return {"ftype": fof.VectorField}
+
+        return {"ftype": fof.ArrayField}
+
+    if isinstance(value, dict):
+        return {"ftype": fof.DictField}
+
+    if isinstance(value, ObjectId):
+        return {"ftype": fof.ObjectIdField}
+
+    raise TypeError(
+        "Cannot infer an appropriate field type for value '%s'" % value
+    )
+
+
+def get_embedded_document_fields(value):
+    return [field for name, field in value._fields.items() if name != "_cls"]
+
+
+def _get_list_value_type(value):
+    if isinstance(value, bool):
+        return fof.BooleanField
+
+    if isinstance(value, numbers.Integral):
+        return fof.IntField
+
+    if isinstance(value, numbers.Number):
+        return fof.FloatField
+
+    if isinstance(value, six.string_types):
+        return fof.StringField
+
+    if isinstance(value, ObjectId):
+        return fof.ObjectIdField
+
+    if isinstance(value, foed.BaseEmbeddedDocument):
+        return fof.EmbeddedDocumentField
+
+    if isinstance(value, datetime):
+        return fof.DateTimeField
+
+    if isinstance(value, date):
+        return fof.DateField
+
+    return None
+
+
+def _merge_implied_fields(implied_fields):
```

Comment:
`_merge_implied_fields()` merges a list of implied field kwargs recursively. It has not been thoroughly tested...